lib/semantics/semantics.cc

```diff
@@ -0,0 +1,99 @@
+// Copyright (c) 2018, NVIDIA CORPORATION.  All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "semantics.h"
+#include "mod-file.h"
+#include "resolve-labels.h"
+#include "resolve-names.h"
+#include "rewrite-parse-tree.h"
+#include "scope.h"
+#include "symbol.h"
+
+namespace Fortran::semantics {
+
+static void DoDumpSymbols(std::ostream &, const Scope &, int indent = 0);
+static void PutIndent(std::ostream &, int indent);
+
+Semantics &Semantics::set_searchDirectories(
+    const std::vector<std::string> &directories) {
+  for (auto directory : directories) {
+    directories_.push_back(directory);
+  }
+  return *this;
+}
+
+Semantics &Semantics::set_moduleDirectory(const std::string &directory) {
+  moduleDirectory_ = directory;
+  directories_.insert(directories_.begin(), directory);
+  return *this;
+}
+
+bool Semantics::Perform(parser::Program &program) {
+  ValidateLabels(messages_, program);
+  if (AnyFatalError()) {
+    return false;
+  }
+  ResolveNames(messages_, globalScope_, program, directories_);
+  if (AnyFatalError()) {
+    return false;
+  }
+  RewriteParseTree(messages_, globalScope_, program);
+  if (AnyFatalError()) {
```

Comment:
I believe AnyFatalError() is superlinear, no? Something like O(k*m_1 + (k-1)*m_2 + ... + 1*m_k) where k is the number of semantic passes and m_i are the number of error messages for pass i that are not fatal. ValidateLabels() originally built its own list of messages and threaded back a bool to avoid that.

Reply:
Don't complicate designs to gain insignificant speedups.